drivers/counter/counter_neorv32_gptmr.c

```diff
@@ -0,0 +1,278 @@
+	}
+}
+
+static int neorv32_gptmr_init(const struct device *dev)
+{
+	const struct neorv32_gptmr_config *config = dev->config;
+	uint32_t features;
+	uint32_t ctrl;
+	int err;
+
+	if (!device_is_ready(config->syscon)) {
+		LOG_ERR("syscon device not ready");
+		return -EINVAL;
+	}
+
+	err = syscon_read_reg(config->syscon, NEORV32_SYSINFO_SOC, &features);
+	if (err < 0) {
+		LOG_ERR("failed to determine implemented features (err %d)", err);
+		return -EIO;
+	}
+
+	if ((features & NEORV32_SYSINFO_SOC_IO_GPTMR) == 0) {
+		LOG_ERR("neorv32 gptmr not supported");
+		return -ENODEV;
+	}
+
+	/* Stop timer, set prescaler, clear any pending interrupt */
+	ctrl = FIELD_PREP(NEORV32_GPTMR_CTRL_PRSC, config->prescaler) | NEORV32_GPTMR_CTRL_IRQ_CLR;
+	neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl);
+
+	config->irq_config_func();
+
+	return 0;
+}
+
+static DEVICE_API(counter, neorv32_gptmr_driver_api) = {
+	.start = neorv32_gptmr_start,
+	.stop = neorv32_gptmr_stop,
+	.get_value = neorv32_gptmr_get_value,
+	.set_top_value = neorv32_gptmr_set_top_value,
+	.get_pending_int = neorv32_gptmr_get_pending_int,
+	.get_top_value = neorv32_gptmr_get_top_value,
+	.get_freq = neorv32_gptmr_get_freq,
+};
+
+#define COUNTER_NEORV32_GPTMR_INIT(n)                                                              \
+	static void neorv32_gptmr_config_func_##n(void)                                            \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(n), DT_INST_IRQ(n, priority), neorv32_gptmr_isr,          \
+			    DEVICE_DT_INST_GET(n), 0);                                             \
+		irq_enable(DT_INST_IRQN(n));                                                       \
+	}                                                                                          \
+                                                                                                   \
+	static struct neorv32_gptmr_data neorv32_gptmr_data_##n;                                   \
+                                                                                                   \
+	static struct neorv32_gptmr_config neorv32_gptmr_config_##n = {                            \
+		.info = {                                                                          \
+				.max_top_value = UINT32_MAX,                                       \
+				.freq = 0U,                                                        \
+				.flags = COUNTER_CONFIG_INFO_COUNT_UP,                             \
+				.channels = 0,                                                     \
+		},                                                                                 \
+		.syscon = DEVICE_DT_GET(DT_INST_PHANDLE(n, syscon)),                               \
+		.base = DT_INST_REG_ADDR(n),                                                       \
+		.prescaler = DT_INST_ENUM_IDX(n, prescaler),                                       \
```

Comment:
It might be possible to use `DT_INST_PROP(n, prescaler)` to avoid the need of a [lookup table](https://github.com/zephyrproject-rtos/zephyr/pull/89140/files#diff-fd4054350b9ae665e30b93b96fb8efdc0a29a345db5970fd51bfdfd22836f4b4R171).

Reply:
The code needs both the value (for the actual calculation) and the index (for writing to the register), so I could either store both or use a lookup table. I chose the latter.

Reply:
My bad I missed that point.